lib/index.ts

```diff
@@ -526,57 +531,75 @@ export class Worker {
 			onMessageEventHandler,
 		);
 
-		const [defaultTypeContracts, nonTypeContracts] = _.partition(
-			Object.values(contracts),
-			(contract) => {
-				return contract.type.split('@')[0] === 'type';
-			},
-		);
+		// Upsert worker contracts on new version
+		await this.kernel.runMigration(
+			logContext,
+			packageName,
+			packageVersion,
+			async (_context) => {
+				const [defaultTypeContracts, nonTypeContracts] = _.partition(
+					Object.values(contracts),
+					(contract) => {
+						return contract.type.split('@')[0] === 'type';
+					},
+				);
 
-		const [defaultLoopContracts, defaultContracts] = _.partition(
-			Object.values(nonTypeContracts),
-			(contract) => {
-				return contract.type.split('@')[0] === 'loop';
-			},
-		);
+				const [defaultLoopContracts, defaultContracts] = _.partition(
+					Object.values(nonTypeContracts),
+					(contract) => {
+						return contract.type.split('@')[0] === 'loop';
+					},
+				);
 
-		// Insert type contracts as prerequisite
-		await Promise.all(defaultTypeContracts.map(checkThenReplaceContract));
+				// Insert type contracts as prerequisite
+				await Promise.all(defaultTypeContracts.map(checkThenReplaceContract));
 
-		// Insert loop contracts as prerequisite
-		await Promise.all(defaultLoopContracts.map(checkThenReplaceContract));
+				// Insert loop contracts as prerequisite
+				await Promise.all(defaultLoopContracts.map(checkThenReplaceContract));
 
-		// Insert other worker contracts
-		const contractsToSkip: string[] = [];
-		if (environment.isProduction() && !environment.isCI()) {
-			contractsToSkip.push('user-guest');
-		}
-		const defaultContractsToLoad = _.values(defaultContracts).filter(
-			(contract: ContractDefinition) => {
-				return !contractsToSkip.includes(contract.slug);
+				// Insert other worker contracts
+				const contractsToSkip: string[] = [];
+				if (environment.isProduction() && !environment.isCI()) {
+					contractsToSkip.push('user-guest');
+				}
+				const defaultContractsToLoad = _.values(defaultContracts).filter(
+					(contract: ContractDefinition) => {
+						return !contractsToSkip.includes(contract.slug);
+					},
+				);
+				await Promise.all(defaultContractsToLoad.map(checkThenReplaceContract));
+
+				await Promise.all(
+					actions.map((action) => {
+						return checkThenReplaceContract(action.contract);
```

Comment:
I expected all migrations to be run inside the transaction provided by `_context` which opens with the `select for update`. 

Looking at the implementation of `checkThenReplaceContract` it looks like it would imply a lot of changes since that method uses `commit` and others that are not transactional, but anyway, the question is shouldn't all migrations be transactional ( atomic )?

Reply:
I had initially looked at updating all methods to support being executed within a transaction like you're describing, but it requires a lot of changes that aren't required to get what I'm trying to achieve with this PR. I can see some value in coming back at some point if we have more regular updates on the types of contracts being loaded in on boot, but I'm not convinced its worth it at the moment. My main goal here is to reduce the number of instances bootstrapping worker and plugin contracts into the db from all instances across the entire fleet to just one.

> I think it would help to decide to know what happens if there's an error during the migration; let's say the migration for a contract on a plugin fails. What would happen with the changes on this PR and what would happen with the current version? I think if we keep the current behavior, then it would be ok.

I believe in both cases the init would fail just causing the backend bootstrap to error out. Data-wise, the only real difference should be that a new record would be inserted into the migrations table for the plugin (in your example) with a version of '0.0.0' if its the first time the plugins migration was attempted. I'll double-check by throwing an error during plugin migrations.

Reply:
Throwing an error during plugin migration callback executions seems to have resulted in what I expected - whatever upserts were executed before the error remained in the database while anything after the error was never created. The new record with version `0.0.0` in the migrations table was also created as I mentioned above, but this is meant to be a sane default place holder anyway so it being there doesn't cause problems.